engine/realtimelogger.h

```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <concurrentqueue.h>
+
+#include <fmt/core.h>
+#include <thread>
+
+namespace pong
+{
+
+static constexpr int MAX_LOG_MESSAGE_SIZE = 512;
+
+struct LogData
+{
+    int mLevel { 0 };
+    int mSequenceNum { 0 };
+    char mMessage[MAX_LOG_MESSAGE_SIZE] {};
+};
+
+class RealTimeLogger
+{
+private:
+    static constexpr int MAX_LOGS = 100;
+
+    std::thread mThread {};
+    moodycamel::ConcurrentQueue<LogData> mLogQueue { MAX_LOGS };
+    bool mAlive { true };
+
+    void Run();
+
+public:
+    RealTimeLogger();
+    ~RealTimeLogger();
+
+    template<typename... Args>
+    void Log(int level, const char *format, Args... args)
+    {
+        LogData logData;
+        logData.mLevel = level;
+        logData.mSequenceNum = Logger::GetNextLogNumber();
+        fmt::format_to_n(logData.mMessage, MAX_LOG_MESSAGE_SIZE, format, args...);
```

Comment:
spdlog does not have an interface to get the final log message as a string or char*. So I have to use it eventually in the RT logger to get the timestamp and level formatted. My solution was to use fmt which spdlog uses anyway. Format the general message now and store that in the LogData. Then when we really need to log the sequence number get prepended to the message and everything is forwarded to an spdlog call.

I did attempt to store the variadic arguments themselves to LogData so it can be passed to spdlog when ready. But that came out to be much uglier, and it is likely not possible entirely.

Reply:
Nice, this seems totally reasonable to me. This function seems realtime safe to my eye 👍